init/setup-rwdev.sh

```diff
@@ -2,24 +2,51 @@
 
 set -e
 
-if [ -e /dev/xvdb ] ; then
+dev=/dev/xvdb
+
+if [ -e "$dev" ] ; then
     # The private /dev/xvdb device is present.
 
     # check if private.img (xvdb) is empty - all zeros
-    private_size_512=`blockdev --getsz /dev/xvdb`
-    if dd if=/dev/zero bs=512 count=$private_size_512 2>/dev/null | diff /dev/xvdb - >/dev/null; then
+    private_size_512=`blockdev --getsz "$dev"`
+    if dd if=/dev/zero bs=512 count="$private_size_512" 2>/dev/null | diff "$dev" - >/dev/null; then
         # the device is empty, create filesystem
-        echo "Virgin boot of the VM: creating private.img filesystem" >&2
-        mkfs.ext4 -m 0 -q /dev/xvdb || exit 1
+        echo "Virgin boot of the VM: creating private.img filesystem on $dev" >&2
```

Comment:
I REALLY detest this.  What if the private device is 50GB?  It will take FOREVER to boot.  I have a few of that, wondering why it was slow.

What should be done is a blkid check.  Say, the device label does not match what is expected from this boot?  Then nuke it.  That's what should be done.  That would be destructive to existing VMs, so perhaps a better check would be to read the first megabyte (or first 64 KB?) and nothing more (or maybe the last 1 MB / 64 KB?).  If the first megabyte of the device is zeroes, that is FOR SURE an indication that the device is empty.

100% overkill to read the device completely, especially considering that virgin devices are going to be allocated from zeroes — sparse, remember? — not actual existing data on disk.

Of course, once LVM / ZFS support is in place, that theory should be tested.  At least in ZFS I know that zvol devices contain 100% zeroes when the device is empty.  I just tested the theory.